2747.feature: Customise fixture test ID by declaring an attr `pytest_id` in the object itself.

fgmacedo opened the pull request:
Customise the string used in a test ID for a certain fixture value by declaring an attribute `pytest_id` in the object itself.

Here's a test that exemplifies the proposal:

``` python

    def test_idmaker_custom_instance_id(self):
        from _pytest.python import idmaker

        class Cheese(object):
            def __init__(self, name):
                self.pytest_id = name

        a = Cheese('gouda')
        b = Cheese('edam')

        result = idmaker(("a", "b"), [pytest.param(a, b)])
        assert result == ["gouda-edam"]
```

The main motivator for this feature request/proposal is to give the opportunity to friendly classes and libs to customize their objects to play nice as `fixtures`. As an example and first candidate, I've open-sourced a custom enum type that can be used as choices field for Django, but their `str` is implemented as `str(self.value)`, so there's no way to get a nice `repr` when using this enum with `pytest.mark.parametrize`.

Example usage: https://github.com/loggi/python-choicesenum/pull/2/files

nicoddemus commented on doc/en/fixture.rst:
```diff
@@ -492,12 +492,16 @@ with ``--collect-only`` will show the generated IDs.
 
+   class Cheese(object):
+       def __init__(self, name)
+           self.pytest_id = name
```
Not sure about the name, as a constant perhaps it should be `PYTEST_ID`?

nicoddemus replied:
@RonnyPfannschmidt @The-Compiler what do you guys think?

RonnyPfannschmidt replied:
at first glance, this is a general no go - this is pretty much a hidden structual dependency

fgmacedo commented:
Thanks @nicoddemus! I think that It will be implemented as properties, like this: https://github.com/loggi/python-choicesenum/pull/2/files#diff-e0e2682b7f60f0bd9508c6a493919747R29

fgmacedo commented:
Thanks for your review @RonnyPfannschmidt. The main point that I miss in `pytest` fixtures parametrization is a way to allow the fixture to give a meaningfull id by itself. The current mechanics requires that the IDs should be given explicity for each fixture usage, or you should pass a callable to `ids`.

Why not allow that friendly classes, or classes/libs that want to support being used as fixtures, to format themselves their IDs?  If you've a suggestion about other approach to accomplish this, I'll be glad to know.

For now, I'm unable to give a nice ID when I use my custom Enum class as fixture, because there's no way in `pytest` as a third party library to give custom IDs.

Example, given a specialized custom enum class:

``` python
class UserStatus(ChoicesEnum):
    UNDEFINED = None
    PENDING = 1
    ACTIVE = 2
    INACTIVE = 3
    DELETED = 4
```

This class overrides `__str__` and `__repr__`, so:

```python
assert str(UserStatus.ACTIVE) == '2'
assert repr(UserStatus.ACTIVE) == 'UserStatus(2).ACTIVE'
```

If used as fixture:

``` python
@pytest.mark.parametrize('user_status', UserStatus)
def test_user_should_allow_status_change(user, user_status):
    pass
```

The output of testcases will be something like this:

``` python
test_user_should_allow_status_change[None]
test_user_should_allow_status_change[1]
test_user_should_allow_status_change[2]
test_user_should_allow_status_change[3]
test_user_should_allow_status_change[4]
```

I want the output to be like this:

``` python
test_user_should_allow_status_change[UserStatus.UNDEFINED]
test_user_should_allow_status_change[UserStatus.PENDING]
test_user_should_allow_status_change[UserStatus.ACTIVE]
test_user_should_allow_status_change[UserStatus.INACTIVE]
test_user_should_allow_status_change[UserStatus.DELETED]
```

The end user of my custom Enum class as seing this output can write a custom function and pass It to the `ids` arg. But why? Why I cannot give my own ID to my class?

This could be solved using this proposal, by implementing a property `pytest_id` (or something else, that will be then forced to be converted to `ascii` using the methods that `pytest` already have.

``` python
class UserStatus(ChoicesEnum):
    UNDEFINED = None
    PENDING = 1
    ACTIVE = 2
    INACTIVE = 3
    DELETED = 4

    @property
    def pytest_id(self):
        return "%s.%s" % (self.__class__.__name__, self._name_, )
```

Thanks for your time readind this. Best regards!

RonnyPfannschmidt commented:
@fgmacedo for correct customization of own types we already have a hook provided that is perfect to put custom ids in place without needing a hidden dependency 

https://docs.pytest.org/en/latest/writing_plugins.html#_pytest.hookspec.pytest_make_parametrize_id

fgmacedo commented:
Hi @RonnyPfannschmidt, I can't implement this hook on the `choicesenum` library. Only the end-user can implemented (in their conftest). I want to ship `choicesenum` with built-in support to be used as a Pytest fixture. How to accomplish this?

RonnyPfannschmidt commented:
@fgmacedo all you need is a module in your package called pytest_plugin which implements that hook

then add the setuptools entrypoint and any users installing your lib will have pytest support via setuptools

fgmacedo commented:
Wow! Nice @RonnyPfannschmidt. I'll try. Thanks for your support!

fgmacedo commented:
Hi @RonnyPfannschmidt, @nicoddemus. I've played with the option to turn my lib into a `pytest` plugin.

It actually worked, thanks!

But I'm a bit frustrated with the amount of work required to only give my class a nice repr when used as a fixture. So I'll share my thoughts with you in hope that someday this got easier.

The side-effects and efforts required:

1. The need to know how to write a `pytest` plugin;
1. The need to know how to write and test `pytest_make_parametrize_id` hook;
1. The need to know how to handle/disable loading others plugins like `pytest-django` that are trying to load and blowing away exceptions while you're trying to test your `pytest_make_parametrize_id` hook with `testdir.runpytest` (WIP since `testdir.runpytest('-v', '-p no:django')` don't worked);
1. My lib `choicesenum` now appears on the list of installed `pytest plugins`, with is not what I want since it's not intended to be a plugin, I only want to my `ChoicesEnum` to have a nice display when used as a fixture:
``` 
    Django settings: tests.settings (from environment variable)
    rootdir: /home/fernando/projects/choicesenum, inifile: pytest.ini
    plugins: sugar-0.9.0, flake8-0.8.1, django-3.1.2, cov-2.5.1, **choicesenum-0.1.4**
```
1. With my lib installed,  there's one more hit on the hook `pytest_make_parametrize_id` for each fixture value for every testcase, even if these tests don't use my class as fixture. So if this is the pattern for doing the `write your custom fixture ID this way`, we're introducing a lot of more function calls, instead of a simple `attribute lookup`.

So, if you can reconsider, I still think that an `attribute lookup` is much simpler than all this stuff that are required for now.

Python does this all time and it's called `protocol`: Define a few methods and attributes that if present, is supposed to work in a documented way, and allow things to integrate and play well, see [data model](https://docs.python.org/3/reference/datamodel.html) for examples.

You could call this as `fixture ID protocol`, and search for something like `__pytest_id__` into fixtures to retrieve custom IDs.

Thank you guys, best regards!

RonnyPfannschmidt commented:
i see what you mean, please open a discussion on the mailing list

nicoddemus commented:
Sorry guys for joining so late.

I agree with @RonnyPfannschmidt that a magic attribute might be a problem down the road. I also agree with the points made by @fgmacedo about using a well defined protocol function, which could even receive the same arguments as `pytest_make_parametrize_id` (not entirely sure). `pytest_make_parametrize_id` default implementation might then be `obj.__pytest_id__()` if it exists.

If you don't want to open a discussion about this on the mailing list right now, @fgmacedo please feel free to open a new issue so we can clearly mark it as a proposal. If more people end up wanting the same feature they can find the proposal and manifest their opinions there. 

I'm closing this for now then. Again thanks @fgmacedo for your work and discussions on this! 👍
